streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1986,53 +1986,47 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
   final def mergeWith[R1 <: R, E1 >: E, O2, O3](
     that: ZStream[R1, E1, O2],
     strategy: TerminationStrategy = TerminationStrategy.Both
-  )(l: O => O3, r: O2 => O3): ZStream[R1, E1, O3] = {
-    type Candidate[E, O] = Fiber[Nothing, Take[E, O]]
-    type Loser           = Either[Candidate[E, O], Candidate[E1, O2]]
-
-    def race(left: Pull[R, E, O], right: Pull[R1, E1, O2]): URIO[R1, (Take[E1, O3], Loser)] =
-      Take
-        .fromPull(left)
-        .raceWith[R1, Nothing, Nothing, Take[E1, O2], (Take[E1, O3], Loser)](Take.fromPull(right))(
-          (exit, right) => ZIO.done(exit).map(take => (take.map(l), Right(right))),
-          (exit, left) => ZIO.done(exit).map(take => (take.map(r), Left(left)))
-        )
+  )(l: O => O3, r: O2 => O3): ZStream[R1, E1, O3] =
+    ZStream {
+      import TerminationStrategy.{ Left => L, Right => R, Either => E }
 
-    def termination[E, O](left: Boolean, right: Boolean, mapper: O => O3)(fiber: Candidate[E, O]) =
-      strategy match {
-        case TerminationStrategy.Either =>
-          fiber.interrupt.as(Exit.fail(None))
-        case TerminationStrategy.Left if left =>
-          fiber.interrupt.as(Exit.fail(None))
-        case TerminationStrategy.Right if right =>
-          fiber.interrupt.as(Exit.fail(None))
-        case _ =>
-          fiber.join.map(_.map(mapper).exit.map((_, (left, right, None))))
+      for {
+        handoff <- ZStream.Handoff.make[Take[E1, O3]].toManaged_
+        done    <- RefM.makeManaged[Option[Boolean]](None)
+        chunksL <- self.process
+        chunksR <- that.process
+        handler = (pull: Pull[R1, E1, O3], terminate: Boolean) =>
+          done.get.flatMap {
+            case Some(true) =>
+              ZIO.succeedNow(false)
+            case _ =>
+              pull.run.flatMap { exit =>
+                done.modify {
+                  (_, exit.fold(c => Left(Cause.sequenceCauseOption(c)), Right(_))) match {
+                    case (state @ Some(true), _) =>
+                      ZIO.succeedNow((false, state))
+                    case (state, Right(chunk)) =>
+                      handoff.offer(Take.chunk(chunk)).as((true, state))
+                    case (_, Left(Some(cause))) =>
+                      handoff.offer(Take.halt(cause)).as((false, Some(true)))
+                    case (option, Left(None)) if terminate || option.isDefined =>
+                      handoff.offer(Take.end).as((false, Some(true)))
+                    case (None, Left(None)) =>
+                      ZIO.succeedNow((false, Some(false)))
+                  }
+                }
+              }
+          }.repeat(Schedule.doWhileEquals(true))
+        _ <- handler(chunksL.map(_.map(l)), List(L, E).contains(strategy)).fork.toManaged_
+        _ <- handler(chunksR.map(_.map(r)), List(R, E).contains(strategy)).fork.toManaged_
```

Comment:
```suggestion
        _ <- handler(chunksL.map(_.map(l)), List(L, E).contains(strategy)).forkManaged
        _ <- handler(chunksR.map(_.map(r)), List(R, E).contains(strategy)).forkManaged
```

Tests pass with this change, and we get the benefit of knowing that the running streams are definitely interrupted before their ZManaged finalizers run.

Reply:
Ah `zipWithLatest` is failing. I'll check why...

Reply:
@iravid yes, I had that one failing with `forkManaged`

Reply:
Pulling was hanging for some reason. Both fibers were stuck and the test was timing out

Reply:
Looks like a bug in Handoff, I think. Digging into it.

Reply:
`zipWithLatest` isn't hanging on JS and it does hang on the JVM even with an unbounded queue.

Reply:
I think `Handoff` (or `Queue`) are unrelated here. It's hanging on `pull.run` (or at least was when I tried yesterday before making more suggested changes) which hasn't offered anything yet at that point to the queue/handoff.

Reply:
Ah it might be something about the TestClock and the fact that `forkManaged` uses `forkDaemon`.

Reply:
Hmm no. Changing forkDaemon in ZManaged#fork back to fork doesn't fix it either.

Reply:
Well it's definitely something with the TestClock. The test hangs in the schedule's sleep.

Reply:
Are you looking into it or is this something Adam should check?

Reply:
Yes, found the problem. We're just using it wrong. The `adjust` call happens concurrently with the fibers sleeping in the stream. Checking how best to fix it now. Basically the same thing as `assertWithChunkCoordination` fixes.

Reply:
Oh I must have not saved the file between running the tests. Changing `forkDaemon -> fork` in `ZManaged#fork` does fix the problem.

Now I'm pretty sure it's because the TestClock doesn't see the producer fibers suspending.

For now @luis3m let's change this from `.fork.toManaged_` to `.fork.interruptible.toManaged(_.interrupt)`. This works correctly.

Reply:
@iravid done